### PR TITLE
[IOTDB-615] Use binary rather than string in insert plan

### PR DESCRIPTION
**Pull request by SilverNarcissus**
Use binary rather than string in insert plan

### CODE REVIEW

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java [hunk 1]
```diff
@@ -168,32 +178,47 @@ private static void insertRecords() throws IoTDBConnectionException, BatchExecut
...
   }
-
   /**
+<<<<<<< HEAD
+   * insert a batch data of one device, each batch contains multiple timestamps with values of
+   * sensors
+   * <p>
+   * a RowBatch example:
+   * <p>
+   * device1 time s1, s2, s3 1,   1,  1,  1 2,   2,  2,  2 3,   3,  3,  3
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply example/session/src/main/java/org/apache/iotdb/SessionExample.java:218]: Fixed~

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java [hunk 2]
```diff
@@ -168,32 +178,47 @@ private static void insertRecords() throws IoTDBConnectionException, BatchExecut
...
+   * <p>
+   * Users need to control the count of RowBatch and write a batch when it reaches the maxBatchSize
+=======
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply example/session/src/main/java/org/apache/iotdb/SessionExample.java:221]: Fixed~

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java [hunk 3]
```diff
@@ -205,6 +230,7 @@ private static void insertRecords() throws IoTDBConnectionException, BatchExecut
    * 3,   3,  3,  3
    *
    * Users need to control the count of Tablet and write a batch when it reaches the maxBatchSize
+>>>>>>> 73dd5eb6e44e2194b650eea72aabcfaf929e5bdc
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply example/session/src/main/java/org/apache/iotdb/SessionExample.java:233]: Fixed~

File: example/session/src/main/java/org/apache/iotdb/SessionExample.java [hunk 4]
```diff
@@ -168,32 +178,47 @@ private static void insertRecords() throws IoTDBConnectionException, BatchExecut
...
         timestamps.clear();
       }
     }
 
-    session.insertRecords(deviceIds, timestamps, measurementsList, valuesList);
+    session.insertRecords(deviceIds, timestamps, measurementsList, typesList, valuesList);
   }
-
   /**
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply example/session/src/main/java/org/apache/iotdb/SessionExample.java:212]: Fixed~

File: server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertPlan.java [hunk 5]
```diff
@@ -174,8 +234,102 @@ public void serializeTo(DataOutputStream stream) throws IOException {
...
+
+  public TSDataType[] getTypes() {
+    return types;
+  }
+
+  public void setTypes(TSDataType[] types) {
+    this.types = types;
+  }
+
+  public void getValues(ByteBuffer buffer) throws QueryProcessException {
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/qp/physical/crud/InsertPlan.java:308]: Sure~

File: server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java [hunk 6]
```diff
@@ -1111,7 +1118,9 @@ public TSStatus insertRecord(TSInsertRecordReq req) {
       plan.setDeviceId(req.getDeviceId());
       plan.setTime(req.getTimestamp());
       plan.setMeasurements(req.getMeasurements().toArray(new String[0]));
-      plan.setValues(req.getValues().toArray(new String[0]));
+      plan.setTypes(new TSDataType[plan.getMeasurements().length]);
+      plan.setValues(new Object[plan.getMeasurements().length]);
+      plan.getValues(req.values);
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java:1123]: Fixed~

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 7]
```diff
@@ -393,16 +359,17 @@ public void insertTablets(Map<String, Tablet> tablets, boolean sorted)
...
+   * your performance, please see insertTablet method
+   * <p>
    * Each row is independent, which could have different deviceId, time, number of measurements
    *
    * @see Session#insertTablet(Tablet)
    */
   public void insertRecords(List<String> deviceIds, List<Long> times,
-      List<List<String>> measurementsList, List<List<String>> valuesList)
+      List<List<String>> measurementsList, List<List<TSDataType>> typesList,
+      List<List<Object>> valuesList)
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply session/src/main/java/org/apache/iotdb/session/Session.java:372]: If user uses wrong type, we will fall at server end. I think we need fall at client to follow the "fall quickly" principle.

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 8]
```diff
@@ -215,66 +218,29 @@ private boolean reconnect() {
...
-   *     time s1, s2, s3
-   *     1,   1,  1,  1
-   *     2,   2,  2,  2
-   *     3,   3,  3,  3
-   *
+   * <p>
+   * a Tablet example:
+   * <p>
+   * device1 time s1, s2, s3 1,   1,  1,  1 2,   2,  2,  2 3,   3,  3,  3
+   * <p>
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply session/src/main/java/org/apache/iotdb/session/Session.java:243]: Sure~

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 9]
```diff
@@ -215,66 +218,29 @@ private boolean reconnect() {
...
+   * insert data in one row, if you want to improve your performance, please use insertRecords
+   * method or insertTablet method
    *
-   * @see Session#insertRecords(List, List, List, List)
+   * @see Session#insertRecords(List, List, List, List, List)
    * @see Session#insertTablet(Tablet)
    */
   public void insertRecord(String deviceId, long time, List<String> measurements,
+      List<TSDataType> types,
       Object... values) throws IoTDBConnectionException, StatementExecutionException {
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply session/src/main/java/org/apache/iotdb/session/Session.java:229]: This method is add for some client using dynamic parameter.

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 10]
```diff
@@ -215,66 +218,29 @@ private boolean reconnect() {
...
   }
 
-  /**
-   * insert data in one row, if you want to improve your performance, please use insertRecords method
-   * or insertTablet method
-   *
-   * @see Session#insertRecords(List, List, List, List)
-   * @see Session#insertTablet(Tablet)
-   */
-  public void insertRecord(String deviceId, long time, List<String> measurements,
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply session/src/main/java/org/apache/iotdb/session/Session.java:241]: Thanks~ Could you give a function signature here?

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 11]
```diff
@@ -215,66 +218,29 @@ private boolean reconnect() {
...
-   * or insertTablet method
+   * insert data in one row, if you want to improve your performance, please use insertRecords
+   * method or insertTablet method
    *
-   * @see Session#insertRecords(List, List, List, List)
+   * @see Session#insertRecords(List, List, List, List, List)
    * @see Session#insertTablet(Tablet)
    */
   public void insertRecord(String deviceId, long time, List<String> measurements,
+      List<TSDataType> types,
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply session/src/main/java/org/apache/iotdb/session/Session.java:228]: If user uses wrong type, we will fall at server end. I think we need fall at client to follow the "fall quickly" principle.

File: service-rpc/rpc-changelist.md [hunk 12]
```diff
@@ -121,3 +121,35 @@ Last Updated on 2019-10-27 by Lei Rui.
 | Add required i64 statementId in TSExecuteStatementReq        | Yuan Tian |
 | Add required binary time, required list<binary> valueList, required list<binary> bitmapList and remove required binary values, required i32 rowCount in TSQueryDataSet| Yuan Tian |
 | Add optional i32 fetchSize in TSExecuteStatementReq,<br />Add optional TSQueryDataSet in TSExecuteStatementResp| liutaohua |
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply service-rpc/rpc-changelist.md:124]: Sure~

File: example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java [hunk 13]
```diff
@@ -89,8 +90,37 @@ private void insert(String data) throws IoTDBConnectionException, StatementExecu
     String device = dataArray[0];
     long time = Long.parseLong(dataArray[1]);
     List<String> measurements = Arrays.asList(dataArray[2].split(":"));
-    List<String> values = Arrays.asList(dataArray[3].split(":"));
-    session.insertRecord(device, time, measurements, values);
+    List<TSDataType> types = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java:93]: Fixed~

File: server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java [hunk 14]
```diff
@@ -1066,15 +1066,22 @@ public TSExecuteBatchStatementResp insertRecords(TSInsertRecordsReq req) {
...
+        plan.getValues(req.valuesList.get(i));
+        TSStatus status = checkAuthority(plan, req.getSessionId());
+        if (status != null) {
+          resp.addToStatusList(status);
+        } else {
+          resp.addToStatusList(executePlan(plan));
+        }
+      }
+      catch (Exception e) {
+        logger.error("meet error when insert in batch", e);
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply server/src/main/java/org/apache/iotdb/db/service/TSServiceImpl.java:1084]: The status has been set to the resp at 
```
resp.addToStatusList(executePlan(plan));
```

File: session/src/main/java/org/apache/iotdb/session/Session.java [hunk 15]
```diff
@@ -434,26 +408,103 @@ public void insertRecords(List<String> deviceIds, List<Long> times,
...
+          ReadWriteIOUtils.write((Long) values.get(i), buffer);
+          break;
+        case FLOAT:
+          ReadWriteIOUtils.write((Float) values.get(i), buffer);
+          break;
+        case DOUBLE:
+          ReadWriteIOUtils.write((Double) values.get(i), buffer);
+          break;
+        case TEXT:
+          ReadWriteIOUtils.write(new Binary((String) values.get(i)), buffer);
```
[the remark this answers is not in the document]
Comment (SilverNarcissus) [reply session/src/main/java/org/apache/iotdb/session/Session.java:467]: Fixed~

File: service-rpc/rpc-changelist.md [hunk 16]
```diff
@@ -121,3 +121,34 @@ Last Updated on 2019-10-27 by Lei Rui.
...
+| Add struct TSQueryNonAlignDataSet                            | Haonan Hou|
+
+## 3. Update
+
+| Latest Changes                                               | Related Committers     |
+| ------------------------------------------------------------ | ---------------------- |
+| Replace TS_SessionHandles with SessionIds, TSOperationHandle with queryIds  | Tian Jiang  |
+| Add optional TSQueryNonAlignDataSet in TSExecuteStatementResp, TSFetchResultsResp and required bool isAlign in TSFetchResultsReq | Haonan Hou |
+| Rename TSStatusType to TSStatus   | Jialin Qiao   |
+| Remove sessionId in TSExecuteBatchStatementResp   | Jialin Qiao   |
```
Comment (HTHou) [inline service-rpc/rpc-changelist.md:153]: I think this is a part of the out-of-date content. It's better to remove it.

File: service-rpc/rpc-changelist.md [hunk 17]
```diff
@@ -121,3 +121,34 @@ Last Updated on 2019-10-27 by Lei Rui.
...
+| Use TsDataType and binary rather than string in TSInsertInBatchReq and TSInsertReq  | Kaifeng Xue  |
```
Comment (HTHou) [inline service-rpc/rpc-changelist.md:154]: Move this line to line 53.

Comment (HTHou) [inline service-rpc/rpc-changelist.md:154]: Move this line to line 53.

File: example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java [hunk 18]
```diff
@@ -89,8 +90,37 @@ private void insert(String data) throws IoTDBConnectionException, StatementExecu
...
+    for(String type : dataArray[3].split(":")){
```
Comment (HTHou) [inline example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java:94]:
```suggestion
    for (String type : dataArray[3].split(":")) {
```

File: example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java [hunk 19]
```diff
@@ -89,8 +90,37 @@ private void insert(String data) throws IoTDBConnectionException, StatementExecu
...
+      types.add(TSDataType.valueOf(type));
+    }
+
+    List<Object> values = new ArrayList<>();
+    String[] valuesStr = dataArray[4].split(":");
+    for(int i = 0; i < valuesStr.length; i++){
```
Comment (HTHou) [inline example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java:100]:
```suggestion
    for (int i = 0; i < valuesStr.length; i++) {
```

File: example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java [hunk 20]
```diff
@@ -89,8 +90,37 @@ private void insert(String data) throws IoTDBConnectionException, StatementExecu
...
+      switch (types.get(i)){
```
Comment (HTHou) [inline example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java:101]:
```suggestion
      switch (types.get(i)) {
```

File: flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java [hunk 21]
```diff
@@ -49,12 +52,37 @@ public Event serialize(Map<String,String> tuple) {
             measurements = Arrays.asList(tuple.get(fieldMeasurements).split(separator));
         }
 
-        List<String> values = null;
-        if (tuple.get(fieldValues) != null) {
-            values = Arrays.asList(tuple.get(fieldValues).split(separator));
+        List<TSDataType> types = new ArrayList<>();
+        for(String type : tuple.get(fieldTypes).split(separator)){
```
Comment (HTHou) [inline flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java:56]:
```suggestion
        for (String type : tuple.get(fieldTypes).split(separator)) {
```

File: flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java [hunk 22]
```diff
@@ -49,12 +52,37 @@ public Event serialize(Map<String,String> tuple) {
...
+            types.add(TSDataType.valueOf(type));
         }
 
-        return new Event(device, timestamp, measurements, values);
+        List<Object> values = new ArrayList<>();
+        String[] valuesStr = tuple.get(fieldValues).split(separator);
+        for(int i = 0; i < valuesStr.length; i++){
```
Comment (HTHou) [inline flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java:62]:
```suggestion
        for (int i = 0; i < valuesStr.length; i++) {
```

File: flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java [hunk 23]
```diff
@@ -49,12 +52,37 @@ public Event serialize(Map<String,String> tuple) {
...
+            switch (types.get(i)){
```
Comment (HTHou) [inline flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java:63]:
```suggestion
            switch (types.get(i)) {
```

**Comment by jixuan1989** [comment]
We need to keep compatible with previous versions.. so we'd better to retain the APIs that allow Strings.

By the way, sometimes users may do not know the detailed data type the series is on the server when they write data.

File: service-rpc/rpc-changelist.md (hunk 16 above)
Comment (SilverNarcissus) [reply service-rpc/rpc-changelist.md:153]: Sure~

File: service-rpc/rpc-changelist.md (hunk 17 above)
Comment (SilverNarcissus) [reply service-rpc/rpc-changelist.md:154]: Sure~

File: example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java (hunk 18 above)
Comment (SilverNarcissus) [reply example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java:94]: fixed~

File: example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java (hunk 19 above)
Comment (SilverNarcissus) [reply example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java:100]: fixed~

File: example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java (hunk 20 above)
Comment (SilverNarcissus) [reply example/rocketmq/src/main/java/org/apache/iotdb/rocketmq/RocketMQConsumer.java:101]: fixed~

File: flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java (hunk 21 above)
Comment (SilverNarcissus) [reply flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java:56]: fixed~

Comment (SilverNarcissus) [reply flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java:56]: fixed~

File: flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java (hunk 22 above)
Comment (SilverNarcissus) [reply flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java:62]: fixed~

File: flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java (hunk 23 above)
Comment (SilverNarcissus) [reply flink-iotdb-connector/src/main/java/org/apache/iotdb/flink/DefaultIoTSerializationSchema.java:63]: fixed~

**Comment by SilverNarcissus** [comment]
> We need to keep compatible with previous versions.. so we'd better to retain the APIs that allow Strings.
> 
> By the way, sometimes users may do not know the detailed data type the series is on the server when they write data.

Thanks~  I have add original interface back and test

File: session/src/main/java/org/apache/iotdb/session/Session.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (qiaojialin) [reply session/src/main/java/org/apache/iotdb/session/Session.java:212]: public void insertRecord(String deviceId, long time, List<String> measurements, List<String> values)

**Pull request merged**